File: Prácticas/Deepspace_P3/test/test_p2.rb
#!/usr/bin/env ruby
#encoding:utf-8

require_relative "../lib/Hangar"
require_relative "../lib/EnemyStarShip"
require_relative "../lib/GameUniverse"
require_relative "../lib/Damage"
require_relative "../lib/SpaceStation"
require_relative "../lib/HangarToUI"

include Deepspace

# Class to make the tests
class TestP2
	def self.main
		puts "TEST CLASS: Hangar"
		puts "================================================================================"
		2.times do
			maxElements = rand(10)
			puts "Creating Hangar.new(#{maxElements})"
			hangar_test = Hangar.new(maxElements)
			puts "Result: #{hangar_test}"
			puts "Adding seven different weapons"
			i = 0
			7.times do
				i += 1
				name = "certainWeapon#{i}"
				type = [WeaponType::LASER, WeaponType::MISSILE, WeaponType::PLASMA][rand(3)]
				uses = rand(10)
				weapon = Weapon.new(name, type, uses)
				puts "Trying to add: #{weapon.to_s}"
				if hangar_test.addWeapon(weapon) == false
					puts "\t#{name} cannot be added on iteration #{i}"
				else
					puts "\t#{name} can be added on iteration #{i}"
				end
			end
			puts "Hangar state is: #{hangar_test}"
			puts "Removing two weapons"
			2.times do 
				removed_weapon = hangar_test.removeWeapon(0)
				puts "\tWeapon removed: #{removed_weapon}"
			end
			puts "Hangar state is:"
			puts "\t#{hangar_test}"

			puts "Adding seven different shield boosters"
			i = 0
			7.times do
				i += 1
				name = "certainShieldBooster#{i}"
				boost = rand(9) + rand()
				uses = rand(10)
				shield = ShieldBooster.new(name, boost, uses)
				puts "Trying to add booster: #{shield.to_s}"
				if hangar_test.addShieldBooster(shield) == false
					puts "\t#{name} cannot be added on iteration #{i}"
				else
					puts "\t#{name} can be added on iteration #{i}"
				end
			end
			puts "Hangar state is:"
			puts "\t#{hangar_test}"

			puts "Removing two shields"
			2.times do
				removed_shield = hangar_test.removeShieldBooster(0)
				puts "\tShield removed: #{removed_shield}"
			end
		
			puts "Testing getters:"
			puts "\tWeapons: #{hangar_test.weapons}"
			puts "\tShields: #{hangar_test.shieldBoosters}"
			puts "\tMax elements: #{hangar_test.maxElements}"
			puts "\tSpace available: #{hangar_test.spaceAvailable}"
			puts "\tUI version: #{hangar_test.getUIversion}"
			puts "--------------------------------------------------------------------------------"
		end

		puts "TEST CLASS: Damage"
		puts "================================================================================"

		3.times do
			puts "Creating both types of damage..."
			nWeapons = rand(10)
			nShields = rand(10)
			puts "Creating Damage.newNumericWeapons(#{nWeapons}, #{nShields})"
			numeric_damage_test = Damage.newNumericWeapons(nWeapons, nShields)
			weapons = []
			4.times do
				weapons << [WeaponType::LASER, WeaponType::MISSILE, WeaponType::PLASMA][rand(3)]
			end
			nShields = rand(10)
			puts "Creating Damage.newSpecificWeapons(#{weapons}, #{nShields})"
			type_damage_test = Damage.newSpecificWeapons(weapons, nShields)

			puts "Numeric Damage: \n\t#{numeric_damage_test}"
			puts "Specific damage: \n\t#{type_damage_test}"
			puts ""

			puts "Creating a copy of Numeric Damage"
			copy_test = Damage.newCopy(numeric_damage_test)
			puts "Copy Damage: \n\t#{copy_test}"
			puts ""

			puts "Creating UI version"
			ui_damage_test = numeric_damage_test.getUIversion
			puts "UI version: #{ui_damage_test}"
			puts ""

			puts "Adjusting the Specific Damage"
			srink_damage_test = type_damage_test.adjust([WeaponType::LASER, WeaponType::MISSILE], ["test", "this does nothing"])
			puts "\tAdjusted Damage: #{srink_damage_test}"
			puts "\tAdjusted Damage Weapons: #{srink_damage_test.weapons}"
			puts "" 

			puts "Testing hasNoEffect"
			puts "Has no effect: #{numeric_damage_test.hasNoEffect}"
			puts ""

			puts "Testing discardShieldBooster()"
			10.times do
					numeric_damage_test.discardShieldBooster
			end
			puts "Has no effect: #{numeric_damage_test.hasNoEffect}"
			puts ""

			puts "Testing discardWeapon"
			10.times do
					numeric_damage_test.discardWeapon("trash")
			end
			puts "Has no effect: #{numeric_damage_test.hasNoEffect}"
			puts ""
		end

		puts "TEST CLASS: EnemyStarShip"
		puts "================================================================================"
		
		puts "Creating an enemy starship"
		enemy_test = EnemyStarShip.new("enemy tank ship", 1.123, 2.123124, Loot.new(1,2,3,4,5), Damage.newNumericWeapons(3,3))
		puts "Enemy starship: #{enemy_test}"
		puts ""

		puts "Creating a copy of the enemy starship"
		enemy_copy = EnemyStarShip.newCopy(enemy_test)
		puts "Copy: #{enemy_copy}"
		puts ""

		puts "Creating an UI version of the ship"
		enemy_UI = enemy_test.getUIversion
		puts "UI version: #{enemy_UI}"
		puts ""

		puts "Protecting: #{enemy_test.protection}"
		puts ""

		puts "Firing: #{enemy_test.fire}"
		puts ""

		puts "Testing EnemyStarShip.receiveShot()"
		10.times do |power|
				puts "\tAttacking with #{power} power!"
				result = enemy_test.receiveShot(power)
				puts "\tResult: #{result}"
		end

		puts "TEST CLASS: Space Station"
		puts "================================================================================"

		puts "Creating an Space Station"
		space_test = SpaceStation.new("my tanky space station", SuppliesPackage.new(14, 12, 40))
		puts "Space Station: #{space_test}"
		puts ""

		puts "Testing speed"
		puts "Speed: #{space_test.speed}"
		puts "Space station state: #{space_test}"
		puts "Valid state: #{space_test.validState}"
		puts ""

		puts "Testing move"
		space_test.move
		puts "Space station state: #{space_test}"

		puts "Testing receive Hangar"
		space_test.receiveHangar(Hangar.new(5))
		puts "Space station: #{space_test}"
		puts ""

		puts "Testing UI version"
		puts "UI: #{space_test.getUIversion}"
		puts ""


		puts "TEST CLASS: GameUniverse"
		puts "================================================================================"
		puts "Creating a Game universe"
		universe_test = GameUniverse.new
		puts "Game Universe: #{universe_test}"
		puts ""

		puts "Testing getters:"
		puts "\tGame State: #{universe_test.gameState}"
		puts ""

		puts "Have a winner: #{universe_test.haveAWinner}"
		puts ""
	end
end

# Launching the tests
TestP2.main

File: Prácticas/deepspace/src/deepspace/TestP2.java
/**
 * @author Sergio Quijano Rey
 * @file TestP2.java
 * @brief Tests for all the code implemented at the second session
 * */

package deepspace;

import java.util.ArrayList;

class TestP2{
    public static void main(String[] args){
        // Testing hangar
        //======================================================================
        System.out.println("Testing Hangar");
        System.out.println("================================================================================");
       
        // Creating an hangar
        Hangar test_hangar = new Hangar(5);
        System.out.println("The created hangar is " + test_hangar.toString());

        // Creating a copy
        Hangar test_hangar_copy = new Hangar(test_hangar);
        System.out.println("The copy of the previous hangar is " + test_hangar_copy.toString());

        // Testing the UI Version 
        HangarToUI test_hangar_ui = test_hangar.getUIversion();
        System.out.println("The hangar UI Version is " + test_hangar_ui.toString());

        // Testing add weapon and add shield
        Weapon test_weapon = new Weapon("test weapon", WeaponType.LASER, 2);
        ShieldBooster test_shield = new ShieldBooster("test shield", (float)2.3123, 4);

        System.out.println("Adding some weapons and shields");
        for(int i = 0; i < 3; i++){
            if(test_hangar.addWeapon(test_weapon)){
                System.out.println("Weapon added succesfully to test_weapon");
            }else{
                System.out.println("Weapon failed to add to test_weapon");
            }
        }

        for(int i = 0; i < 3; i++){
            if(test_hangar.addShieldBooster(test_shield)){
                System.out.println("ShieldBooster added succesfully to test_weapon");
            }else{
                System.out.println("ShieldBooster failed to add to test_weapon");
            }
        }

        // We show that the copy is different from the original
        System.out.println("State of the original hangar: " + test_hangar.toString());
        System.out.println("State of the copy: " + test_hangar_copy.toString());

        // We remove a few weapons
        System.out.println("Removing two weapons");
        for(int i = 0; i < 2; i++){
            test_hangar.removeWeapon(0);
        }

        System.out.println("Removing two shields");
        for(int i = 0; i < 2; i++){
            test_hangar.removeShieldBooster(0);
        }

        System.out.println("State of the original hangar: " + test_hangar.toString());
        System.out.println("State of the copy: " + test_hangar_copy.toString());

        System.out.println("");
        System.out.println("");

        // Testing Damage
        //======================================================================
        System.out.println("Testing Damage");
        System.out.println("================================================================================");
        ArrayList<WeaponType> weapon_type_list = new ArrayList<WeaponType>(0);
        weapon_type_list.add(WeaponType.LASER);
        weapon_type_list.add(WeaponType.MISSILE);
        Damage test_damage_a = new NumericDamage(3, 2);
        Damage test_damage_b = new SpecificDamage(weapon_type_list, 2);
        System.out.println("First damage: " + test_damage_a.toString());
        System.out.println("Second damage: " + test_damage_b.toString());

        // Testing copies
        Damage test_damage_a_copy = test_damage_a.copy();
        Damage test_damage_b_copy = test_damage_b.copy();
        System.out.println("The copy of the first damage is: " + test_damage_a_copy);
        System.out.println("The copy of the second damage is: " + test_damage_b_copy);

        // Testing the adjust method
        System.out.println("TESTING Damage.adjust()");
        
        ArrayList<Weapon> weapons_to_adjust = new ArrayList<>(0);
        weapons_to_adjust.add(new Weapon("first weapon", WeaponType.LASER, 3));
        
        ArrayList<ShieldBooster> shields_to_adjust = new ArrayList<ShieldBooster>();
        shields_to_adjust.add(new ShieldBooster("shield to adjust", (float)2.3134, 4));        
        shields_to_adjust.add(new ShieldBooster("shield to adjust", (float)2.3134, 4));
 
        Damage adjusted = test_damage_a.adjust(weapons_to_adjust, shields_to_adjust);
        System.out.println("First adjusted damage is " + adjusted.toString());
        adjusted = test_damage_b.adjust(weapons_to_adjust, shields_to_adjust);
        System.out.println("Second adjusted damage is " + adjusted.toString());

        // Testing discard methods
        System.out.println("Discarding some weapons");
        for(int i = 0; i < 4; i++){
            System.out.println("Discarding " + i + " weapon on first damage");     
            test_damage_a.discardWeapon(test_weapon);
            System.out.println("Discarding " + i + " weapon on second damage");
            test_damage_b.discardWeapon(test_weapon);
        }

        System.out.println("Discarding some shields");
        for(int i = 0; i < 4; i++){
            System.out.println("Discarding " + i + " shield on first damage");
            test_damage_a.discardShieldBooster();
            System.out.println("Discarding " + i + " shield on second damage");
            test_damage_b.discardShieldBooster();
        }

        System.out.println("State of the first damage: " + test_damage_a.toString());
        System.out.println("State of the second damage: " + test_damage_b.toString());

        System.out.println("");
        System.out.println("");

        // Testing EnemyStarShip
        //======================================================================
        System.out.println("Testing EnemyStarShip");
        System.out.println("================================================================================");
        Damage test_damage = new NumericDamage(2, 3);
        Loot test_loot = new Loot(1, 2, 3, 4, 5, false, false);
        
        // Testing the constructor
        EnemyStarShip test_enemy = new EnemyStarShip("Enemy Test", (float)1.1, (float)2.2, test_loot, test_damage);
        System.out.println("The testing EnemyStarShip is " + test_enemy.toString());

        // Testing the copy constructor
        EnemyStarShip test_enemy_copy = new EnemyStarShip(test_enemy);
        System.out.println("The copy of the enemy is " + test_enemy_copy.toString());

        // Testing the UI version
        System.out.println("The UI version is " + test_enemy.getUIversion().toString());

        // Testing some shots
        System.out.println("The enemy is going to receive some shots");
        for(int i = 0; i < 10; i++){
            ShotResult shotResult = test_enemy.receiveShot(i);
            System.out.println("Receiving " + i + " power shot results in " + shotResult);
        }
        
        System.out.println("");
        System.out.println("");
    } 
}
